qiskit_experiments/curve_analysis/fit_models.py

```diff
@@ -0,0 +1,199 @@
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""Fit models that are used for curve fitting."""
+
+from abc import ABC, abstractmethod
+from typing import Callable, List, Optional
+
+import numpy as np
+
+from qiskit_experiments.exceptions import AnalysisError
+
+
+class FitModel(ABC):
+    """Base class of fit models.
+
+    This is a function-like object that implements a fit model as a ``__call__`` magic method,
+    thus it behaves like a Python function that the SciPy curve_fit solver accepts.
+    Note that the fit function there only accepts variadic arguments.
+
+    This class ties together the fit function and associated parameter names to
+    perform correct parameter mapping among multiple objective functions with different signatures,
+    in which some parameters may be excluded from the fitting when they are fixed.
+    """
+
+    def __init__(
+        self,
+        fit_functions: List[Callable],
+        signatures: List[List[str]],
+        fit_models: Optional[List[str]] = None,
+        fixed_parameters: Optional[List[str]] = None,
+    ):
+        """Create new fit model.
+
+        Args:
+            fit_functions: List of callables that defines fit function of a single series.
+            signatures: List of parameter names of a single series.
+            fit_models: List of string representation of fit functions.
+            fixed_parameters: List of parameter names that are fixed in the fit.
+
+        Raises:
+            AnalysisError: When ``fit_functions`` and ``signatures`` don't match.
+        """
+        if len(fit_functions) != len(signatures):
+            raise AnalysisError("Different numbers of fit_functions and signatures are given.")
+
+        self._fit_functions = fit_functions
+        self._signatures = signatures
+        self._fit_models = fit_models or [None for _ in range(len(fit_functions))]
+
+        if not fixed_parameters:
+            fixed_parameters = []
+        self._fixed_params = {p: None for p in fixed_parameters}
+
+        # Create signature of this fit model, i.e. this will be signature of scipy fit function.
+        # The curves comprising this model may have different signature.
+        # The signature of this fit model is union of parameters in all curves.
+        union_params = []
+        for signature in signatures:
+            for parameter in signature:
+                if parameter not in union_params and parameter not in fixed_parameters:
+                    union_params.append(parameter)
+        self._uniton_params = union_params
```

Comment:
This will be breaking API change because @Parameters entry in the curve analysis returns a list of fit parameter values in this order. Of course there is another list for parameter name so user can always convert it into dict with `dict(zip(param_names, popts))`. However some external user may have already built own package to track this vector for monitoring, for example.